ksql-serde/src/main/java/io/confluent/ksql/serde/avro/AvroDataTranslator.java

```diff
@@ -0,0 +1,172 @@
+    }
+    return new GenericRow(columns);
+  }
+
+  @Override
+  public Struct toConnectRow(final GenericRow genericRow) {
+    final List<Object> columns = new LinkedList<>();
+    for (int i = 0; i < genericRow.getColumns().size(); i++) {
+      columns.add(
+          replaceSchema(
+              avroCompatibleSchema.fields().get(i).schema(),
+              genericRow.getColumns().get(i)));
+    }
+    return innerTranslator.toConnectRow(new GenericRow(columns));
+  }
+
+  private static class TypeNameGenerator {
+    private static final String DELIMITER = "_";
+    private static final String DEFAULT_SCHEMA_NAME_BASE = "KSQLDefaultSchemaName";
+
+    static final String MAP_KEY_NAME = "MapKey";
+    static final String MAP_VALUE_NAME = "MapValue";
+
+    private Iterable<String> names;
+
+    TypeNameGenerator() {
+      this(ImmutableList.of(DEFAULT_SCHEMA_NAME_BASE));
+    }
+
+    private TypeNameGenerator(Iterable<String> names) {
+      this.names = names;
+    }
+
+    TypeNameGenerator with(String name) {
+      return new TypeNameGenerator(Iterables.concat(names, ImmutableList.of(name)));
+    }
+
+    public String name() {
+      return String.join(DELIMITER, names);
+    }
+  }
+
+  private String avroCompatibleFieldName(final Field field) {
+    // Currently the only incompatible field names expected are fully qualified
+    // column identifiers. Once quoted identifier support is introduced we will
+    // need to implement something more generic here.
+    return field.name().replace(".", "_");
+  }
+
+  private Schema buildAvroCompatibleSchema(final Schema schema,
+                                           final TypeNameGenerator typeNameGenerator) {
+    final SchemaBuilder schemaBuilder;
+    switch (schema.type()) {
+      default:
+        return schema;
+      case STRUCT:
+        schemaBuilder = SchemaBuilder.struct();
+        if (schema.name() == null) {
+          schemaBuilder.name(typeNameGenerator.name());
+        }
+        for (final Field f : schema.fields()) {
+          schemaBuilder.field(
+              avroCompatibleFieldName(f),
+              buildAvroCompatibleSchema(f.schema(), typeNameGenerator.with(f.name())));
+        }
+        break;
+      case ARRAY:
+        schemaBuilder = SchemaBuilder.array(
+            buildAvroCompatibleSchema(schema.valueSchema(), typeNameGenerator));
+        break;
+      case MAP:
+        schemaBuilder = SchemaBuilder.map(
+            buildAvroCompatibleSchema(schema.keySchema(),
+                typeNameGenerator.with(TypeNameGenerator.MAP_KEY_NAME)),
+            buildAvroCompatibleSchema(schema.valueSchema(),
+                typeNameGenerator.with(TypeNameGenerator.MAP_VALUE_NAME)));
+        break;
+    }
+    if (schema.isOptional()) {
+      schemaBuilder.optional();
+    }
+    return schemaBuilder.build();
+  }
+
+  @SuppressWarnings("unchecked")
+  private Object replaceSchema(Schema schema, Object object) {
```

Comment:
nit: `final`
FYI: You can change your setting in IDEA to highlight non-final params etc.